triggers/service/src/main/scala/com/digitalasset/daml/lf/engine/trigger/ServiceMain.scala

```diff
@@ -64,12 +64,49 @@ case class LedgerConfig(
 object Server {
 
   sealed trait Message
-  private final case class StartFailed(cause: Throwable) extends Message
-  private final case class Started(binding: ServerBinding) extends Message
+
   final case class GetServerBinding(replyTo: ActorRef[ServerBinding]) extends Message
   final case object Stop extends Message
 
-  private def addDar(compiledPackages: MutableCompiledPackages, dar: Dar[(PackageId, Package)]) = {
+  private final case class StartFailed(cause: Throwable) extends Message
+  private final case class Started(binding: ServerBinding) extends Message
+
+  final case class TriggerStarting(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message])
+      extends Message
+  final case class TriggerStarted(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message])
+      extends Message
+  final case class TriggerInitializationFailure(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message],
+      cause: String
+  ) extends Message
+  final case class TriggerRuntimeFailure(
+      triggerId: UUID,
+      jwt: Jwt,
+      runner: ActorRef[TriggerRunner.Message],
+      cause: String
+  ) extends Message
+
+  case class TriggerRunnerWithToken(
+      ref: ActorRef[TriggerRunner.Message],
+      token: Jwt
+  )
+
+  // The server is some kind of "singleton". We need this state at
+  // this scope in that we may bind it into the following closures. We
+  // reinitialize this state, each time we go through 'apply'.
+  private var triggers: Map[UUID, TriggerRunnerWithToken] = Map.empty;
```

Comment:
looks like this didn’t change in this PR but do we really need global mutable state here? why not make this a class and have the state be local to the class?

Reply:
The object level state has been removed. Have moved the state back into `apply` in the interim. Will follow up on bundling it into a class.